kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,196 @@
+              lastSeenCheckpointVersion);
+
+          numDeleted += deleteLogFiles(engine, potentialLogFilesToDelete);
+          potentialLogFilesToDelete.clear();
+
+          // Step 1.2: add `lastCheckpointFiles` to `potentialFileStoDelete` list. This checkpoint
+          // is potential candidate to delete later if we find another checkpoint
+          potentialLogFilesToDelete.addAll(lastSeenCheckpointFiles);
+          lastSeenCheckpointFiles.clear();
+          lastSeenCheckpointVersion = -1;
+        }
+
+        FileStatus nextFile = files.next();
+
+        // Step 2: If the timestamp is earlier than the retention period, stop
+        if (nextFile.getModificationTime() > fileCutOffTime) {
+          if (!potentialLogFilesToDelete.isEmpty()) {
+            logger.info(
+                "{}: Skipping deletion of expired log files {}, because there is no checkpoint "
+                    + "file that indicates that the log files are no longer needed. ",
+                tablePath,
+                potentialLogFilesToDelete.size());
+          }
+          break;
+        }
+
+        if (FileNames.isCommitFile(nextFile.getPath())) {
+          // Step 3: If the file is a delta log file, add it to the `potentialFilesToDelete` list
+          // We can't delete these files until we encounter a checkpoint later that indicates
+          // that the log files are no longer needed.
+          potentialLogFilesToDelete.add(nextFile.getPath());
+        } else if (FileNames.isCheckpointFile(nextFile.getPath())) {
+          // Step 4: If the file is a checkpoint file, add it to the `lastSeenCheckpointFiles`
+          long newLastSeenCheckpointVersion = FileNames.checkpointVersion(nextFile.getPath());
+          checkArgument(
+              lastSeenCheckpointVersion == -1
+                  || newLastSeenCheckpointVersion >= lastSeenCheckpointVersion);
+
+          if (lastSeenCheckpointVersion != -1
+              && newLastSeenCheckpointVersion > lastSeenCheckpointVersion) {
+            // We have found checkpoint file for a new version. This means the files gathered for
+            // the last checkpoint version are not complete (most likely an incomplete multipart
+            // checkpoint). We should delete the files gathered so far and start fresh
+            // last seen checkpoint state
+            logger.info(
+                "{}: Incomplete checkpoint files found at version {}, ignoring the checkpoint"
+                    + " files and adding them to potential log file delete list",
+                tablePath,
+                lastSeenCheckpointVersion);
+            potentialLogFilesToDelete.addAll(lastSeenCheckpointFiles);
+            lastSeenCheckpointFiles.clear();
+          }
+
+          lastSeenCheckpointFiles.add(nextFile.getPath());
+          lastSeenCheckpointVersion = newLastSeenCheckpointVersion;
+        }
+        // TODO: do we need to delete unknown file types?
```

Comment:
cc @prakharjain09

Reply:
Delta-Spark ignores [unknown](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala#L150) files.

Removing the TODO.